### PR TITLE
Introduce generic Rewriter dispatching between specific rewriters.

### PR DESCRIPTION
**Pull request by mgautierfr**
Some js coming from specific website need to be patched to work "offline".

This patching is made using a set of specific regex to "activate" depending of path.

Sadly, there is no test because there is none in wabac.js (the source of the rules set) and
I don't know want/how to test.

We probably have more to import from wabac.js, but this PR is a fix to #149

### CODE REVIEW

**Comment by rgaudin** [comment]
Is this a reimplementation of the JS version in wabac.js? I suppose it was not possible to source it directly.
What's the update process? I think we at least need a comment referencing the source and how to convert from it

**Comment by mgautierfr** [comment]
> Is this a reimplementation of the JS version in wabac.js? I suppose it was not possible to source it directly.

Yes, it is.

> What's the update process? I think we at least need a comment referencing the source and how to convert from it

There is no update process. The simple things have been rewritten by chat-gpt 🙈 (and reviewed by a human). The other have been made by a human directly.

We could create a script which update the rules from the wabac source but it would not be straightforward as the "action" part of the rule is a bit different.
And if the code change, we will have to adapt manual anyway.


---

Last 4 commits are structural improvement but no functional changes.

**Comment by benoit74** [comment]
Is it correct to add a comment in `ds.py` saying that all rules present in this file comes from `https://github.com/webrecorder/wabac.js/commits/main/src/rewrite/dsruleset.js`and have been manually converted from Javascript to Python (I would tell details about ChatGPT, who cares ...)? I would add this at least, as suggest by Renaud. I don't believe we can have a more precise conversion tool / procedure (unfortunately). 

I'm not very comfortable with the rest of the code, but "that's life" 😄

**Comment by rgaudin** [comment]
> Is it correct to add a comment in `ds.py` saying that all rules present in this file comes from `https://github.com/webrecorder/wabac.js/commits/main/src/rewrite/dsruleset.js`and have been manually converted from Javascript to Python (I would tell details about ChatGPT, who cares ...)? I would add this at least, as suggest by Renaud. I don't believe we can have a more precise conversion tool / procedure (unfortunately).

That's what I have in mind ; just a starting point for another maintainer in the future: knowing where it came from, being able to check if there's a conversion bug and being able to convert new ones

**Comment by benoit74** [comment]
Maybe also specify the last commit (from wabac.js) which has been taken into account, so that it is clear what is left to be back-ported

**Review by benoit74 — CHANGES_REQUESTED** [review]
could you please change your PR to target `warc2zim2`, rebase your branch and resolve merge conflict (I thought it would be easier this way, also in order to close the other PR) and add the comment we discussed explaining where this code comes from?

**Comment by benoit74** [comment]
Edit: github already decided to target `warc2zim2` automatically (nice ^^)

**Comment by mgautierfr** [comment]
> add a comment in ds.py saying that all rules present in this file comes from https://github.com/webrecorder/wabac.js/commits/main/src/rewrite/dsruleset.jsand have been manually converted from Javascript to Python (I would tell details about ChatGPT, who cares ...)? 

There was already a comment in the commit message, but now it is also in the code.
(Directly in modifying the commit, no fixup)

File: src/warc2zim/content_rewriting/generic.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        if self.rewrite_mode == "javascript":
+            return self.rewrite_js(opts)
+
+        return ("", self.content)
+
+    def get_rewrite_mode(self, record, mimetype):
+        if getattr(record, "method", "GET") == "POST":
+            return None
+        if mimetype == "text/html":
+            # TODO : Handle header "Accept" == "application/json"
```
Comment (benoit74) [inline src/warc2zim/content_rewriting/generic.py:50]: What does it mean? Do you plan to do it soon? Is there an open issue around what has to be done?

File: src/warc2zim/content_rewriting/ds.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    rx_rules: list[TransformationRule]
+
+
+MAX_BITRATE = 5000000
+
+
+def set_max_bitrate(_opts: dict) -> int:
+    max_bitrate = MAX_BITRATE
+
+    # Extra opts is used in wabac, but we never set it ourself here.
```
Comment (benoit74) [inline src/warc2zim/content_rewriting/ds.py:35]: I'm sorry but I don't get what you mean here. Where do these `extraOpts` come from? The Twitter/Vimeo response? Then why don't we process them just like wabac does?

Maybe you should simply open an issue explaining:
- why we do not set extra_opts (we never did?)
- what could be the driver to decide to set them

File: src/warc2zim/content_rewriting/ds.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
+## THis file is almost a plain translation from js file
```
Comment (benoit74) [inline src/warc2zim/content_rewriting/ds.py:1]: Many typos / wrong English. Corrected version suggestion:
```
## This file is almost a plain translation from a JavaScript file.
## https://github.com/webrecorder/wabac.js/blob/219830cea0a732bdd72ce100fcfd2f3a1c9c3607/src/rewrite/dsruleset.js
## Partly translated by a human, partly by ChatGPT.

## The ruleset should be kept in sync with the original ones, but there is no easy
## automatic process for that at the moment.
```

**Review by benoit74 — CHANGES_REQUESTED** [review]
Almost OK, two important questions (could be tackled in a separate issue if needed) and a suggestion to fix typos.

File: src/warc2zim/content_rewriting/generic.py (hunk 1 above)
Comment (mgautierfr) [reply src/warc2zim/content_rewriting/generic.py:50]: Wabac is a checking if the header `Accept` is `application/json` and then use the json converter even if mimetype is `text/html`.

We don't have a json rewriter (we probably need one) and so I've just left a comment here to keep a trace that we may have something to do here.

> Do you plan to do it soon?

This has to be discussed. Until now, we had implement transformation based on what needed (this is why we agreed to not keep response header for now). But we may change our strategy and reimplement (blindly) the transformation in wabac. I will open a issue.

File: src/warc2zim/content_rewriting/ds.py (hunk 2 above)
Comment (mgautierfr) [reply src/warc2zim/content_rewriting/ds.py:35]: We are two :)

Wabac's `set_max_bitrate` is searching/saving data in `extraOpts`. But I don't get from where it comes from and what to do with it.

File: src/warc2zim/content_rewriting/ds.py (hunk 3 above)
Comment (mgautierfr) [reply src/warc2zim/content_rewriting/ds.py]: Fixed (directly in the commit)

File: src/warc2zim/content_rewriting/generic.py (hunk 1 above)
Comment (benoit74) [reply src/warc2zim/content_rewriting/generic.py:50]: I've open an issue to track the question (at least): https://github.com/openzim/warc2zim/issues/160

File: src/warc2zim/content_rewriting/ds.py (hunk 2 above)
Comment (benoit74) [reply src/warc2zim/content_rewriting/ds.py:35]: I've opened https://github.com/openzim/warc2zim/issues/161 to track the question (at least)

**Pull request merged**